### PR TITLE
Fix, clean up, and generify broadcast[!] over pairs of sparse matrices

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request fixes (#19110), cleans up, and slightly generifies `broadcast[!]` over pairs of sparse matrices.

This pull request does not fundamentally change existing semantics or implementation. Apart from the fix for #19110, changes that reduce cognitive load for wetware interpreters (removal of meta-fu, transformation of variable names and expressions to more enlightening alternatives, insertion of comments) constitute most of the diff. About half the additions derive from the #19110 fix, which unfortunately duplicates part of the existing implementation. (The bulk of that duplication could go away in exchange for some efficiency (both time and memory).)

This pull request represents a first step, not where the code should land. The next step requires settling `map!`/ `broadcast!`'s semantics for sparse matrices, and steps thereafter require determining `broadcast_zpreserving!`'s fate. I will open issues to hash those design questions out.

Best!

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -1504,148 +1504,281 @@ end
...
-                            nzvalB[ptrB] = res
-                            ptrB += 1
-                        end
-                        ptr1 += 1
-                        ptr2 += 1
-                    end
+## Broadcast operations involving two sparse matrices
+function broadcast{Tf}(f::Tf, A::SparseMatrixCSC, B::SparseMatrixCSC)
+    indextypeC = promote_type(eltype(A.rowval), eltype(B.rowval))
+    entrytypeC = Base.Broadcast.promote_eltype_op(f, A, B)
```
Comment (ararslan) [inline base/sparse/sparsematrix.jl]: The `Base.` prefix shouldn't be necessary, right?

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: `promote_eltype_op` is not exported, necessitating the `Base.Broadcast.` qualification? Thanks!

Comment (ararslan) [inline base/sparse/sparsematrix.jl]: Right, you need `Broadcast` but since this is part of Base you shouldn't need `Base`, I think.

Comment (yuyichao) [inline base/sparse/sparsematrix.jl]: This file isn't part of `Base`.

Comment (ararslan) [inline base/sparse/sparsematrix.jl]: Okay, my mistake. Sorry for the noise.

**Comment by Sacha0** [comment]
Fixed / improved a few lines with benefit of what I learned from #19438. Absent objections or requests for time, I plan to merge this Thursday morning PST. Best!

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: `promote_eltype_op` is defined in `abstractarray.jl` in `Base` so you can drop the `Broadcast`

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Nixed the qualification's `Broadcast` part. Thanks!

**Comment by Sacha0** [comment]
Rebased. Planning to merge once CI clears. Best!

**Pull request merged**